crates/prover/src/runner.rs

```diff
@@ -319,218 +304,182 @@ where
-                        match retry_backoff(exponential_backoff.clone(), || async move {
-                            match inner_client.get_soft_batch::<Da::Spec>(l2_height).await {
-                                Ok(Some(soft_batch)) => Ok(soft_batch),
-                                Ok(None) => {
-                                    debug!("Soft Batch: no batch at height {}", l2_height);
-
-                                    // Return a Permanent error so that we exit the retry.
-                                    Err(backoff::Error::Permanent(
-                                        "No soft batch published".to_owned(),
-                                    ))
-                                }
-                                Err(e) => match e.downcast_ref::<JsonrpseeError>() {
-                                    Some(JsonrpseeError::Transport(e)) => {
-                                        let error_msg = format!(
-                                            "Soft Batch: connection error during RPC call: {:?}",
-                                            e
-                                        );
-                                        error!(error_msg);
-                                        Err(backoff::Error::Transient {
-                                            err: error_msg,
-                                            retry_after: None,
-                                        })
-                                    }
-                                    _ => {
-                                        let error_msg = format!(
-                                            "Soft Batch: unknown error from RPC call: {:?}",
-                                            e
-                                        );
-                                        error!(error_msg);
-                                        Err(backoff::Error::Transient {
-                                            err: error_msg,
-                                            retry_after: None,
-                                        })
-                                    }
-                                },
+            let mut sof_soft_confirmations_to_push = vec![];
+            let mut da_block_headers_to_push: Vec<
+                <<Da as DaService>::Spec as DaSpec>::BlockHeader,
+            > = vec![];
+
+            // start fetching blocks from sequencer, when you see a soft batch with l1 height more than end_l1_height, stop
+            // while getting the blocks to all the same ops as full node
+            // after stopping call continue  and look for a new seq_commitment
+            // change the item numbers only after the sync is done so not for every da block
+            loop {
+                let inner_client = &self.sequencer_client;
+                let soft_batch = match retry_backoff(exponential_backoff.clone(), || async move {
+                    match inner_client.get_soft_batch::<Da::Spec>(l2_height).await {
```

Comment:
IMO get_soft_batch deserves it's own method/function, because of how huge this block of code is.

Reply:
IIUC, this can be done in https://github.com/chainwayxyz/citrea/issues/738